### ISSUE TITLE
TypeScript reports errors when union type passed as Body to RouteGeneric since Fastify 4

### ISSUE THREAD

**Issue by svsool** [fastify/fastify#4063]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.1.0

### Node.js version

18.x

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

12.3

### Description

Hi, I came across the following problem after updating to 4.x.

`Body` property of request gets resolved to `unknown` instead of `{ filters?: { ids: string[] } } | null`. Use-case: we have some routes that optionally accept body using ajv's `anyOf` keyword. It used to work in Fastify 3.

Fastify 4:
![image](https://user-images.githubusercontent.com/1452141/174833671-158068b3-7ac9-492d-9dbf-4187b6d3dcbd.png)

vs.

Fastify 3:
![image](https://user-images.githubusercontent.com/1452141/174833367-31be37c0-699b-41cf-8441-9981666071f7.png)

### Steps to Reproduce

This snippet can be used to reproduce the problem:

```
import { fastify } from 'fastify';

const app = fastify();

app.route<{
    Body: { filters?: { ids: string[] } } | null;
}>({
    method: 'DELETE',
    url: '/entities',
    schema: {}, // omitted for clarity
    async handler(request, reply) {
        const filters = request.body?.filters;
    },
});

```

### Expected Behavior

`body` type gets resolved to `{ filters?: { ids: string[] } } | null` instead of unknown.

**Comment by A5rocks** [comment]
I ran into this and investigated a bit (didn't realize someone else had made an issue). Here's what I found:

Firstly, [`body`](https://github.com/fastify/fastify/blob/4361770a1278cc518ddff683c5944dfb1db470c1/types/request.d.ts#L36) is governed by `ResolveFastifyRequestType<TypeProvider, SchemaCompiler, RouteGeneric>["body"]`. In my case, `TypeProvider` is `FastifyTypeProviderDefault`, `SchemaCompiler` is `FastifySchema`, `RouteGeneric` is `{ Body: z.infer<typeof body>, [snip] }` where `body` is a `zod.union([..., ..., ...])`.

This then gets moved into [`ResolveRequestBody`](https://github.com/fastify/fastify/blob/4361770a1278cc518ddff683c5944dfb1db470c1/types/type-provider.d.ts#L34-L35).

https://github.com/fastify/fastify/blob/4361770a1278cc518ddff683c5944dfb1db470c1/types/type-provider.d.ts#L34-L35

See, `keyof <union>` is `never`. This means it goes to `CallTypeProvider<TypeProvider, SchemaCompiler['body']>` which returns `unknown` (as we have not extended `FastifyTypeProviderDefault`).

What is there to do? Well, checking StackOverflow, I find that something like [`type KeysOf<T> = T extends T ? keyof T : never;`](https://stackoverflow.com/a/49402091) can be used for unions. The only problem is that I'm not quite sure why this conditional exists in the first place for `ResolveRequestBody`.

tl;dr

Why is `ResolveRequestBody` conditional on `keyof <body type>`?

If that is intentional, then `type KeysOf<T> = T extends T ? keyof T : never;` would work; `keyof <body type>` -> `KeysOf<<body type>>`.

**Comment by mcollina** [comment]
Thanks for reporting! Would you like to work on a PR with a fix? I'm not sure if your proposed solution is ok, however it mostly depends if it breaks any other tests.

**Comment by A5rocks** [comment]
I believe I fix this in https://github.com/fastify/fastify/pull/4076
